packages/structure/src/lib.rs

```diff
@@ -0,0 +1,237 @@
+use serde_json::{from_str, json, Value};
+use std::env;
+use TOCEntryType::*;
+
+#[macro_export]
+macro_rules! inline_content {
+    ($expr:expr) => {
+        json!({
+            "name": "inline_content",
+            "data": $expr
+        })
+    }
+}
+
+pub struct StructureCounter {
+    headings: Vec<u64>,
+    figures: u64,
+    tables: u64,
+}
+
+#[derive(PartialEq)]
+enum TOCEntryType {
+    Numbered,
+    Unnumbered,
+    Empty,
+}
+
+struct TOCEntry {
+    id: Option<String>,
+    contents: Option<Value>,
+    children: Vec<TOCEntry>,
+    mode: TOCEntryType,
+}
+
+pub struct TOC {
+    table: TOCEntry,
+    max_level: usize,
+}
+
+impl StructureCounter {
+    pub fn new() -> Self {
+        Self {
+            headings: vec![0],
+            figures: 0,
+            tables: 0,
+        }
+    }
+    // Enforce a limit of 256 levels, to avoid creating an extremely long vec from malformed input.
+    pub fn push_heading(&mut self, level: usize) {
+        if level < 1 || level > u8::MAX as usize {
+            return;
+        }
+
+        if level == 1 {
+            self.figures = 0;
+            self.tables = 0;
+        }
+
+        while self.headings.len() < level {
+            self.headings.push(0);
+        }
+
+        self.headings[level - 1] += 1;
+
+        for counter in &mut self.headings[level..] {
+            *counter = 0;
+        }
+    }
+
+    pub fn push_figure(&mut self) {
+        self.figures += 1;
+    }
+
+    pub fn push_table(&mut self) {
+        self.tables += 1;
+    }
+
+    pub fn get_heading(&self) -> String {
+        self.headings
+            .iter()
+            .map(|c| c.to_string())
+            .collect::<Vec<String>>()
+            .join(".")
+            .trim_end_matches(".0")
+            .to_string()
+    }
+
+    pub fn get_figure(&self) -> String {
+        format!("{}.{}", self.headings[0], self.figures)
+    }
+
+    pub fn get_table(&self) -> String {
+        format!("{}.{}", self.headings[0], self.tables)
+    }
+}
+
+impl TOCEntry {
+    fn numbered(id: String, contents: Value) -> Self {
+        TOCEntry {
+            id: Some(id),
+            contents: Some(contents),
+            children: vec![],
+            mode: Numbered,
+        }
+    }
+
+    fn unnumbered(id: String, contents: Value) -> Self {
+        TOCEntry {
+            id: Some(id),
+            contents: Some(contents),
+            children: vec![],
+            mode: Unnumbered,
+        }
+    }
+
+    fn empty() -> Self {
+        TOCEntry {
+            id: None,
+            contents: None,
+            children: vec![],
+            mode: Empty,
+        }
+    }
+}
+
+impl TOC {
+    // Build and return a TOC from the elements in "structure". Currently, only "numbered-heading"
+    // and "unnumbered-heading" are added.
+    pub fn build_from_list(max_level: usize) -> Self {
```

Comment:
Nice! This would be ease to generalize for a more flexible [listings] module in the future if users want to create a table of figures etc. too.